lib/src/source/hosted.dart

```diff
@@ -1100,3 +1109,91 @@ class _RefAndCache {
   @override
   bool operator ==(Object other) => other is _RefAndCache && other.ref == ref;
 }
+
+const checksumHeaderName = 'x-goog-hash';
+
+bool _hasChecksumHeader(Map<String, String> headers) {
+  return headers.containsKey(checksumHeaderName);
+}
+
+/// Adds a checksum validation "tap" to the response stream and returns a
+/// wrapped Stream object, which should be used to consume the incoming data.
+///
+/// As chunks are received, a CRC32C checksum is updated.
+/// Once the download is completed, the final checksum is compared with
+/// the one present in the checksum response header.
+///
+/// Throws [PackageIntegrityException] if anything is wrong with the checksum
+/// or if there is a checksum mismatch.
+Stream<List<int>> _responseStreamWithChecksumValidationTap(
+    http.StreamedResponse response, String fileName) {
+  final hostedCrc32c = _parseCrc32c(response.headers);
+  if (hostedCrc32c == null) {
+    throw PackageIntegrityException(
+        'Package response headers have an invalid or missing CRC32C checksum',
+        url: response.request?.url);
+  }
+
+  return Crc32c.computeByTappingStream(response.stream, handleDone: (crc32c) {
+    log.fine(
+        'Computed CRC32C ($crc32c) for $fileName with hosted CRC32C of ($hostedCrc32c)');
+
+    if (hostedCrc32c != crc32c) {
+      throw PackageIntegrityException(
+          'Package fetched from host has a CRC32C checksum mismatch; Computed checksum ($crc32c) != Hosted checksum ($hostedCrc32c)',
+          url: response.request?.url);
+    }
+  });
+}
+
+/// Parses response headers and returns the package's CRC32C checksum.
+///
+/// In most cases, GCS provides both MD5 and CRC32C checksums in its response
+/// headers. It uses the header name "x-goog-hash" for these values. It has
+/// been documented and observed that GCS will send multiple response headers
+/// with the same "x-goog-hash" token as the key.
+/// https://cloud.google.com/storage/docs/xml-api/reference-headers#xgooghash
+///
+/// Additionally, when the Dart http client encounters multiple response
+/// headers with the same key, it concatenates their values with a comma
+/// before inserting a single item with that key and concatenated value into
+/// its response "headers" Map.
+/// See https://github.com/dart-lang/http/issues/24
+/// https://github.com/dart-lang/http/blob/06649afbb5847dbb0293816ba8348766b116e419/pkgs/http/lib/src/base_response.dart#L29
+int? _parseCrc32c(Map<String, String> responseHeaders) {
+  String? checksums = responseHeaders[checksumHeaderName];
+  if (checksums == null) return null;
+
+  final parts = checksums.split(',');
+  for (final part in parts) {
+    if (part.startsWith('crc32c=')) {
+      final undecoded = part.substring('crc32c='.length);
+
+      try {
+        final bytes = base64.decode(undecoded);
+        return bytesToUint32(bytes);
+      } catch (e) {
```

Comment:
Please avoid catch-alls.
Probably we just need to catch `FormatException` here.

Reply:
Yes, I'll remove the catch-all. I think we'll also need to catch `RangeError`.

Reply:
Nevermind, I see that Errors should be handled differently. https://stackoverflow.com/questions/57000995/throwing-exceptions-in-dart-for-invalid-parameters-e-g-sign-in/57004304#57004304